Talvihoitoreitit

rckopteri opened the pull request:
** Talvihoitoreitit **

***API***
Apin kautta voi lisätä ja päivittää Talvihoitoreittejä.
APIa varten on tehty 
Json schemat.
Json esimerkit vastaa schemoja

***EXCEL import***
Oletus on, että suurin osa Talvihoitoreiteistä lisätän Excelillä. 

***Hallintapaneelin työkalu***
APIa on testattu ja kehitetty Hallintapaneelin työkalun avulla.
Hallintapaneeliin on lisätty myös sivu, josta näkee pyörivän ympäristön tieosoitteet. Tätä tarvitaan, jos esim GC ympäristössä haluaa lisätä Excelin kautta Talvihoitoreitin ja nähdä sen geometrian kartalla.
Talvihoitoreitit eivät siis ota vastaan sellaisia reittejä, joille ei ole Harjassa tukea.

***UI***
UI:lla näkyy reitit ja kalustot. Jokaiselle reitille arvotaan sisääntullessa väri, jota käytetään aina, kun reitti piirretään listaan tai kartalle.

***Yksikkötestit***
Testataan APIa ja perushakuja.

***Uudet talvihoitoluokat***
Tähän asti tieosoitteilla on ollut 10 erilaista talvihoitoluokka. Tässä tulee mukana kolme uutta, jotka kohdistuu vain Huoltoaukoille ja pysäköintialueille. Koordinaatistolla ne on esim isojen valtaväylien P-paikkoja tai vastaavia.

rckopteri committed:
Alustavia talvihoitoreittijuttuja

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

# Conflicts:
#	src/clj/harja/palvelin/integraatiot/api/tyokalut/json_skeemat.clj

rckopteri committed:
Päivitys skeemaan ja esimerkkitiedosto

rckopteri committed:
Helpota hallintapaneelin testityökalujen käyttöoikeuksien lisäämistä

rckopteri committed:
Muutamia muutoksia talvihoitoreittien näyttämiseksi alustavasti kartalla

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

rckopteri committed:
Talvihoitoreittien API endpoint

rckopteri committed:
Lisää talvihoitoreitti api ja endpoint main tiedostoon

rckopteri committed:
Viimeistele talvihoitoreitti skeemaa ja esimerkkiä

rckopteri committed:
Valmistele talvihoitoreittien sql haut

rckopteri committed:
Talvihoitoreittien API endpoint kuntoon

rckopteri committed:
Tee endpoint talvihoitoreittien hakemiseksi ui:lle

rckopteri committed:
Tee työkalu valmiiksi, jolla voi lähettää talvihoitoreittejä APIn kautta

rckopteri committed:
Paranna logituksen tasoa

rckopteri committed:
Ensimmäinen versio talvihoitoreittien UIsta

rckopteri committed:
Lisää talvihoitoreitteihin ulkoinen-id

rckopteri committed:
Tee päivitys ja poisto schemat ja niiden validoinnit

rckopteri committed:
Mahdollista talvihoitoreittien poisto ja päivitys

ja pari ui päivitystä

rckopteri committed:
Korjaa insert flywaymigraatiossa

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

rckopteri committed:
Korjaa migraatiotiedostojen nimet

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1131__.sql

rckopteri committed:
Poista kalustotaulu tietokannasta. Ne on liitetty reitin yhteyteen

rckopteri committed:
Yhdistä talvihoitoreitin kalusto reittitietoihin skeemassa, esimerkeissä ja hallinnan työkaluissa

rckopteri committed:
Kalustot siirrettiin reittitauluun, joten muokataan haut vastaamaan sitä

rckopteri committed:
Nimeä talvihoitoreitti-api tiedosto uusiksi

rckopteri committed:
Nimeä palvelu talvihoitoreitit uusiksi

rckopteri committed:
Nimeä talvihoitoreitit tiedot tiedosto uusiksi

rckopteri committed:
Nimeä talvihoitoreitit näkymä tiedosto uusiksi

rckopteri committed:
Viimeiset muutokset tiedostojen nimeämisen kannalta

rckopteri commented on resources/api/schemas/talvihoitoreitti-kirjaus-request.schema.json:
```diff
@@ -0,0 +1,81 @@
+  "type": "object",
+  "additionalProperties": false,
+  "properties": {
+    "ulkoinen-id": {
```
Vaihda ulkoinen-id tunnisteeseen, tai ota vaikka käyttöön se tunniste, joka on käytössä. Voisi kokeilla tehdä uuden, joka on nerokkaasti nimetty, mutta tunnistaa tunnisteeksi.

rckopteri replied:
Vaihdettu

rckopteri commented on tietokanta/src/main/resources/db/migration/V1_1133__.sql:
```diff
@@ -0,0 +1,36 @@
+    loppuosa            INTEGER,
+    loppuetaisyys       INTEGER,
+    hoitoluokka         INTEGER NOT NULL,
+    pituus              INTEGER, -- metreinä
```
Muuta DECIMAL 10,2 koska pituudet voivat olla muutakin kuin integeriä.
Muuta nimi pituus -> pituus_m

rckopteri replied:
Muutettu

rckopteri commented on tietokanta/src/main/resources/db/migration/V1_1133__.sql:
```diff
@@ -0,0 +1,36 @@
+CREATE UNIQUE INDEX talvihoitoreitti_ulkoinen_id_urakka_id_uindex
+    ON public.talvihoitoreitti (ulkoinen_id, urakka_id);
+
+CREATE TABLE IF NOT EXISTS talvihoitoreitti_reitti
```
Nimeä taulu -> talvihoitoreitti_sijainti

rckopteri replied:
Muutettu

rckopteri committed:
Testataanko lähteekö Kapsi kartta toimimaan gc ympäristössä

rckopteri committed:
Verrataan vain osaa hostista

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

rckopteri committed:
Päivitä talvihoitoreitin skeemoja

rckopteri committed:
Mahdollista talvihoitoreittiexcelpohjan lataaminen

rckopteri committed:
Siirrä talvihoitoreitin kalustot omaan tauluunsa

rckopteri committed:
Päivitä hallinnan talvihoitoreittityökalu toimimaan uudemmalla skeemalla

jossa kalustot on sijaintien sisällä

rckopteri committed:
Päivitä talvihoitoreitti näkymä toimimaan uusien map avainten kanssa

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1133__.sql

rckopteri committed:
Siirrä talvihoitoreittien päivitysfunktiot oikeaan tiedostoon

rckopteri committed:
Mahdollista geometrialtaan leikkaavien talvihoitoreittien etsiminen

rckopteri committed:
Lue talvihoitoreitit excelistä

rckopteri committed:
Siivoa turhia logituksia

rckopteri committed:
Lisää talvihoitoreitti API:lle lisää testejä

rckopteri committed:
Vielä kierros yksikkötestien parantamiseksi

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

# Conflicts:
#	test/clj/harja/palvelin/main_test.clj
#	tietokanta/src/main/resources/db/migration/V1_1134__.sql

rckopteri committed:
Lisää puuttuva kirjasto

rckopteri committed:
Päivitä esimerkit vastaamaan skeemaa

rckopteri committed:
Pieni ui säätö

rckopteri committed:
Ota roolit excelin käyttöoikeudet käyttöön talvihoitoreiteissä

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

rckopteri committed:
Varmistus vielä, että roolit excel oli tallennettu

jarnovayrynen commented on src/clj/harja/palvelin/palvelut/laadunseuranta/talvihoitoreitit_excel.clj:
```diff
@@ -0,0 +1,140 @@
+               (and
+                 ;; Annetaan hieman vapauksia kenttien nimille
+                 (or
+                   (= "Reitin nimi*" (first rivi))
```
Tässä lähtee toistumaan tämä "litania", eli funktioon laittaisin tän reitin nimen tarkistuksen. Funktion sisällähän se voi or:ien lisäksi olla settikin, tyyliin (boolean (#("Reitin nimi*" "Reitin nimi" Reitti") rivi)) niin sitten tuolla logiikan sisällä saadaan kutsu yhdelle nimelle, tyyliin 

```clojure
(or 
  (reitin-nimi? (first rivi))
  (tie-sarake? (first rivi))) 


```

rckopteri replied:
Hyvä parannus. Toteutettu

jarnovayrynen commented on src/clj/harja/palvelin/palvelut/laadunseuranta/talvihoitoreitit_excel.clj:
```diff
@@ -0,0 +1,140 @@
+    ;; Poistetaan rivi kokonaan, mikäli nimikenttä on nil. Eli oletetaan että rivillä ei ole
+    ;; annettu muutenkaan mitään asiaan liittyvää tietoa vaan rivi liittyy otsikointiin tms.
+    (fn [rivi]
+      (if (nil? (second rivi))
```
tää vois olla whenillä, joka palauttaa epätodessa aina nil.

```clojure
(when-not (nil? (second rivi)) 
  (do stuff..)
```

rckopteri replied:
Korjattu

solita-staskila replied:
> Korjattu

![Alt Text](https://cdn.betterttv.net/emote/6154a741b63cc97ee6d3eff5/3x.webp)

jarnovayrynen commented on src/clj/harja/palvelin/palvelut/laadunseuranta/talvihoitoreitit_palvelu.clj:
```diff
@@ -0,0 +1,187 @@
+            [harja.palvelin.palvelut.laadunseuranta.talvihoitoreitit-excel :as t-excel]))
+
+(defn hae-urakan-talvihoitoreitit [db user {:keys [urakka-id]}]
+  ;; TODO: HOX muuta roolit excel
```
Jos on jo muutettu (löytyy diffistä), niin todo pois?

rckopteri replied:
Hyvä löytö. Tämä lipsahti minulta kokonaan. Todo poistettu ja oikea oikeustarkistus otettu käyttöön.

jarnovayrynen commented on src/clj/harja/palvelin/palvelut/laadunseuranta/talvihoitoreitit_palvelu.clj:
```diff
@@ -0,0 +1,187 @@
+                                       ;; Hae kaikki kalustot
+                                       reitin-kalustot (conj (flatten (map (fn [r] (:kalustot r)) reitit)))
+                                       ;; Groupataan reitin kalustot kalustotyypin mukaan
+                                       groupatut-kalustot (group-by :kalustotyyppi reitin-kalustot)
```
On jo vähän makuasia, mutta ite laittaisin "ryhmitellyt-kalustot" tähän :D vaikka helppo ymmärtää että groupatut tulee ekana mieleen.

rckopteri replied:
👍

jarnovayrynen commented on src/clj/harja/palvelin/palvelut/laadunseuranta/talvihoitoreitit_palvelu.clj:
```diff
@@ -0,0 +1,187 @@
+
+(defn vastaanota-excel [db request]
+  (log/debug "vastaanota-excel :: request" request)
+  (oikeudet/vaadi-kirjoitusoikeus oikeudet/urakat-paikkaukset-paikkauskohteetkustannukset
```
tähän jäänyt väärä oikeustarkistus?

rckopteri replied:
Kyllä! Hyvä huomio. Nyt korjattu

jarnovayrynen commented on src/clj/harja/palvelin/palvelut/laadunseuranta/talvihoitoreitit_palvelu.clj:
```diff
@@ -0,0 +1,187 @@
+  (let [urakka-id (Integer/parseInt (get (:params request) "urakka-id"))
+        kayttaja (:kayttaja request)]
+    ;; Tarkistetaan, että kutsussa on mukana urakka ja kayttaja
+    (if (and (not (nil? urakka-id))
```
meniskö tää ihan vaan 
```clojure
(if (and urakka-id kayttaja)
``` 
eli tuokin tekee saman "not nil" tarkistuksen, mutta helpompi lukea.

rckopteri replied:
Meneehän se. 👍

jarnovayrynen commented on src/clj/harja/palvelin/palvelut/laadunseuranta/talvihoitoreitit_palvelu.clj:
```diff
@@ -0,0 +1,187 @@
+
+                        ;; Jos talvihoitoreitti löytyy tietokannasta, niin päivitetään
+                        _ (when (and talvihoitoreitti-db (empty? tieosoite-virheet) (empty? leikkaavat-geometriat))
+                            (talvihoitoreitit-q/paivita-talvihoitoreitti-tietokantaan db t urakka-id (:id kayttaja))
```
täällä on useampi kirjoitusoperaatio, mutten löytänyt että nämä oilisivat transaktiossa? Eli pitäis kaikki kirjoitusoperaatiot kääriä transaktioon.

rckopteri replied:
Hyvä idea. Wrappasin doseqin jokaisen yksittäisen tuloksen transaktion sisään.

jarnovayrynen reviewed (changes requested):
Laittelin kommentteja, tärkeimpänä ehkä yksi puuttuva transaktio, ja joitakin pienempiä ajatuksia joita voit itse puntaroida, esim nimeämiset, tai ehtolausekkeen muodot.

rckopteri committed:
Refaktoroitu katselmointikommenttien perusteella

rckopteri committed:
Lisätään vielä cypress-testit talvihoitoreiteille

rckopteri committed:
Muuta testidata takaisin

solita-staskila commented on src/cljc/harja/ui/kartta/infopaneelin_sisalto.cljc:
```diff
@@ -843,6 +843,13 @@
 
+(defmethod infopaneeli-skeema :talvihoitoreitit [talvihoitoreitti]
+  {:tyyppi :talvihoitoreitit
+   :jarjesta-fn (constantly false) #_ :id ;; Järjestetään useampi valittu listalla
```
Onko tässä  jäänyt kesken jotain?

rckopteri replied:
Joo. Tämä on vähä tämmöstä mun omaa lisäilyä. Näistä ei ole speksiä. Otin kommentit pois ja jätin silti näytille.

solita-staskila commented on src/cljc/harja/ui/kartta/infopaneelin_sisalto.cljc:
```diff
@@ -843,6 +843,13 @@
+(defmethod infopaneeli-skeema :talvihoitoreitit [talvihoitoreitti]
+  {:tyyppi :talvihoitoreitit
+   :jarjesta-fn (constantly false) #_ :id ;; Järjestetään useampi valittu listalla
+   :otsikko (str "Talvihoitoreitti: " (:nimi (:infopaneelin-tiedot talvihoitoreitti))) ;; Laita tähän vaikka päivämäärä
```
Ja tässä?

solita-staskila commented on src/cljs/harja/tiedot/hallinta/tyokalut/talvihoitoreitti_tyokalu.cljs:
```diff
@@ -0,0 +1,251 @@
+  [
+   {:id 34 :nimi "Ivalon MHU testiurakka (uusi) (aseta sopimusid 19)"}
+   {:id 35 :nimi "Oulun MHU 2019-2024 (Aseta sopimusid 42)"}
+   ])
```
Näitä voisi formatoida oikein clojureksi, eli ei jätetä noita roikkuvia sulkeita

solita-staskila commented on src/cljs/harja/tiedot/hallinta/tyokalut/talvihoitoreitti_tyokalu.cljs:
```diff
@@ -0,0 +1,251 @@
+    10 "K2"
+    11 "K"))
+
+(defn- luo-aika-paivamaarasta [paivamaara aika]
```
Ei taida olla käytössä tämä?

rckopteri replied:
Ei näimmä enää. Hyvä bongaus!

solita-staskila commented on src/cljs/harja/tiedot/hallinta/tyokalut/talvihoitoreitti_tyokalu.cljs:
```diff
@@ -0,0 +1,251 @@
+
+
+(def valittu-urakka (atom nil))
+(def valittu-hallintayksikko (atom nil))
```
Käytetäänkö mihinkään?

rckopteri replied:
Ei taideta enää käyttää. Otin pois

solita-staskila commented on src/clj/harja/palvelin/palvelut/api_jarjestelmatunnukset.clj:
```diff
@@ -66,6 +66,13 @@
                                                                   :id id})))))
   (hae-jarjestelmatunnuksen-lisaoikeudet db user {:kayttaja-id kayttaja-id}))
 
+(defn tallenna-kayttajalle-kirjoitusoikeus [db user {:keys [oikeudet kayttaja-id]}]
```
Tässä taitaa olla yksi turha destruktoitu muuttuja/parametri

rckopteri replied:
Jep. Otin pois

solita-staskila commented on src/cljs/harja/tiedot/hallinta/tyokalut/talvihoitoreitti_tyokalu.cljs:
```diff
@@ -0,0 +1,251 @@
+  Muokkaa
+  (process-event [{talvihoitoreitti :talvihoitoreitti} app]
+    (let [;; Tarkista hallintayksikko
+          talvihoitoreitti (if (and (not= (get-in app [:talvihoitoreitti :valittu-hallintayksikko]) (:valittu-hallintayksikko talvihoitoreitti)))
```
Turha and?

solita-staskila commented on src/cljs/harja/tiedot/hallinta/tyokalut/talvihoitoreitti_tyokalu.cljs:
```diff
@@ -0,0 +1,251 @@
+        (assoc :talvihoitoreitti talvihoitoreitti))))
+
+  Laheta
+  (process-event [{talvihoitoreitti :talvihoitoreitti} app]
```
-> ```_ app```

solita-staskila commented on src/cljs/harja/tiedot/hallinta/tyokalut/talvihoitoreitti_tyokalu.cljs:
```diff
@@ -0,0 +1,251 @@
+                             :accept :json}
+                     ;; Lähetä talvihoitoreitti rajapintaan
+                     vastaus (<! (http/put (str "api/urakat/" urakkaid "/talvihoitoreitti") params))
+                     _ (println "Laheta :: vastaus" (pr-str vastaus))]
```
Tämän println voi poistaa, tai vaihtaa logiksi jos halutaan info tasolla logittaa

solita-staskila commented on src/cljs/harja/tiedot/hallinta/tyokalut/talvihoitoreitti_tyokalu.cljs:
```diff
@@ -0,0 +1,251 @@
+      (assoc app :lahetys-kaynnissa? true)))
+
+  LahetysOnnistui
+  (process-event [{vastaus :vastaus} app]
```
-> ```_ app``` 

Nämä on vaan ehdotuksia, saa tuo vastaus destruktointi tuossa ollakin jos näissä halutaan kertoa mitä sieltä mahdollisesti  tulee takaisin tai jos tätä halutaan joskus käyttää/logittaa. Mutta yleensä näitä turhia parametreja ei tarvitse definata jos niillä ei mitään tehdä

solita-staskila commented on src/cljs/harja/tiedot/hallinta/tyokalut/talvihoitoreitti_tyokalu.cljs:
```diff
@@ -0,0 +1,251 @@
+
+  HaeHallintayksikonUrakatEpaonnistui
+  (process-event [{vastaus :vastaus} app]
+    (js/console.log "HaeHallintayksikonUrakatEpaonnistui :: vastaus" (pr-str vastaus))
```
Täällä aika paljon console logitusta, halutaanko näitä kaikkea täällä pitää? Vaikka onkin hallinnan puolen työkalu

rckopteri replied:
Näistä iso osa on vaan copypastea toteumatyökalun puolelta. Nämä on pelkästän vain meidän kehityskäyttöön, niin koodin hieronnalla ei saada vastinetta asiakkaan rahoille. En aio siivota näitä kommentteja pois tai refaktoroida tätä, ellen satu tekemään tuohon työkaluun jossain vaiheessa isommin muutoksia.

solita-staskila commented on src/cljs/harja/tiedot/urakka/laadunseuranta/talvihoitoreitit_tiedot.cljs:
```diff
@@ -0,0 +1,230 @@
+            [harja.tyokalut.tuck :as tuck-apurit]
+            [harja.tyokalut.yleiset :as yleiset-tyokalut]
+            [harja.tiedot.navigaatio :as nav]
+            [harja.ui.kartta.ikonit :refer [sijainti-ikoni pinni-ikoni nuoli-ikoni]]
```
Pari turhaa referenssiä

rckopteri replied:
Siivosin pois.

solita-staskila commented on src/cljs/harja/tiedot/urakka/laadunseuranta/talvihoitoreitit_tiedot.cljs:
```diff
@@ -0,0 +1,230 @@
+                            @valitut-kohteet-atom)
+          talvihoitoreitit (keep (fn [reitti] (when (contains? valitut-kohteet (:id reitti)) reitti)) @karttataso-talvihoitoreitit)
+          sijainnit (map (fn [rivi] (map :sijainti (:reitit rivi))) talvihoitoreitit)
+          extentit (harja.geo/extent-monelle (flatten sijainnit))
```
Lisää harja.geo requireen

solita-staskila commented on src/cljs/harja/tiedot/urakka/laadunseuranta/talvihoitoreitit_tiedot.cljs:
```diff
@@ -0,0 +1,230 @@
+
+;; Talvihoitoreittien excel-tiedoston lataamisessa käyttäjän informointiin tarkoitettu modal
+(defn tulos-modal [leikkaavat-virheet muut-virheet paivitetyt onnistuneet otsikko]
+  (let [otsikkotyyli (if (or (not (empty? onnistuneet)) (not (empty? paivitetyt))) :onnistunut :virhe)]
```
Öh, sori kun viilailen vähän pilkkua mutta not empty kahdesti.  Tässä olisi parempi tapa jotain 

`(not (and (empty? onnistuneet) (empty? paivitetyt)))`

Tai vaikka 

`(not (every? empty? [onnistuneet paivitetyt]))`

rckopteri replied:
Mä tykkään tuosta not emptystä. Helppo lukea, kaikki ymmärtää heti, mistä on kyse. Mutta clojure ehdottaa itse, että käytetään seq funkkaria, not emptyn tilalta, niin vaihdoin siihen.

solita-staskila commented on src/cljs/harja/tiedot/urakka/laadunseuranta/talvihoitoreitit_tiedot.cljs:
```diff
@@ -0,0 +1,230 @@
+    (modal/nayta!
+      {:otsikko otsikko
+       :otsikko-tyyli otsikkotyyli}
+      (when (or (not (empty? leikkaavat-virheet)) (not (empty? muut-virheet)) (not (empty? paivitetyt)) (not (empty? onnistuneet)))
```
Oliskohan tässäkin mielummin vaikka:
`(not (every? empty? [leikkaavat-virheet muut-virheet paivitetyt onnistuneet]))`

Tuo not empty on tupla negaatio vai miksikähän sitä kutsutaan

rckopteri replied:
Vaihdoin tämän: (some seq [leikkaavat-virheet muut-virheet paivitetyt onnistuneet])

solita-staskila commented on src/cljs/harja/tiedot/urakka/laadunseuranta/talvihoitoreitit_tiedot.cljs:
```diff
@@ -0,0 +1,230 @@
+       :otsikko-tyyli otsikkotyyli}
+      (when (or (not (empty? leikkaavat-virheet)) (not (empty? muut-virheet)) (not (empty? paivitetyt)) (not (empty? onnistuneet)))
+        [:div
+         (when (not (empty? leikkaavat-virheet))
```
Näissä lopuissa not empty statementeissa taitaa toimia esim:
`(when (seq leikkaavat-virheet) ..`

solita-staskila commented on src/cljs/harja/tiedot/urakka/laadunseuranta/talvihoitoreitit_tiedot.cljs:
```diff
@@ -0,0 +1,230 @@
+                 ^{:key (gensym)}
+                 [:li rivi])]]))
+         (when (not (empty? paivitetyt))
+           (do
```
Redundant do

solita-staskila commented on src/cljs/harja/tiedot/urakka/laadunseuranta/talvihoitoreitit_tiedot.cljs:
```diff
@@ -0,0 +1,230 @@
+                   ^{:key (gensym)}
+                   [:li rivi])])]))
+         (when (not (empty? onnistuneet))
+           (do
```
Redundant do

solita-staskila commented on src/cljs/harja/tiedot/urakka/laadunseuranta/talvihoitoreitit_tiedot.cljs:
```diff
@@ -0,0 +1,230 @@
+
+;; Hae talvihoitoreitit urakalle
+(defn hae-talvihoitoreitit [app]
+  (let []
```
Letissä ei tehdä mitään

rckopteri replied:
Otin pois

solita-staskila commented on src/cljs/harja/tiedot/urakka/laadunseuranta/talvihoitoreitit_tiedot.cljs:
```diff
@@ -0,0 +1,230 @@
+                    vastaus)]
+
+      ;; Jos talvihoitoreittejä löytyy, resetoi kartta
+      (do
```
Tässäkin tätä do ei tarvitse? kö?

solita-staskila commented on src/cljs/harja/tiedot/urakka/laadunseuranta/talvihoitoreitit_tiedot.cljs:
```diff
@@ -0,0 +1,230 @@
+  (process-event [{:keys [vastaus]} app]
+    (let [_ (js/console.log "TiedostoLadattu :: vastaus:" (pr-str vastaus))
+          leikkaavat-virheet (keep #(:leikkaavat % ) (filter #(= (:virheet %)) (:virheet vastaus)))
+          muut-virheet (keep #(get % :virheet) (filter #(= (get-in % [:virheet])) (:virheet vastaus)))
```
Tässä muut-virheet muuttujassa nyt tuon filtterin sisällä jotain logiikan kanssa väärin? Verrataan = operaattorilla itseään. Eli onko tuo aina true? Jotain tässä nyt on väärin, tai muuten vaan turhaa koodia. En ole tätä testannut kirjoitushetkellä, luen pelkästään koodia läpi

rckopteri replied:
Jepa. Kyllä se toimii kuten pitääkin. Se palauttaa true, jos sieltä löytyy lista, eikä nil. Vaihdoin tämän seq funktioon, niin on helpompi lukea.

solita-staskila commented on src/cljs/harja/tiedot/urakka/laadunseuranta/talvihoitoreitit_tiedot.cljs:
```diff
@@ -0,0 +1,230 @@
+          (hae-talvihoitoreitit app))
+
+        ;; Tieosoitevirheet
+        (or (not (empty? leikkaavat-virheet)) (not (empty? muut-virheet)))
```
Toimisikohan tässä taas: 
`(some seq [leikkaavat-virheet muut-virheet])`

solita-staskila commented on src/cljs/harja/views/hallinta/tyokalut/talvihoitoreitit_tyokalu_nakyma.cljs:
```diff
@@ -0,0 +1,135 @@
+            [harja.tiedot.hallintayksikot :as hal]
+            [harja.tiedot.hallinta.tyokalut.talvihoitoreitti-tyokalu :as tiedot])
+
+  (:require-macros [cljs.core.async.macros :refer [go]]))
```
Turhia macro importteja. tuck.coressa turhia referenssejä

solita-staskila commented on src/cljs/harja/views/hallinta/tyokalut/talvihoitoreitit_tyokalu_nakyma.cljs:
```diff
@@ -0,0 +1,135 @@
+  (if (oikeudet/voi-kirjoittaa? oikeudet/hallinta-toteumatyokalu)
+    [:div
+     [debug/debug app]
+     (when (not (empty? (:oikeudet-urakoihin app)))
```
(when (seq (:oikeudet-urakoihin app))

solita-staskila commented on src/cljs/harja/views/hallinta/tyokalut/talvihoitoreitit_tyokalu_nakyma.cljs:
```diff
@@ -0,0 +1,135 @@
+     [:div
+      ;; Näytetään mahdollisuus lisätä oikeudet urakkaan vain, jos siihen ei vielä ole oikeuksia
+      (if (and (get-in app [:talvihoitoreitti :valittu-urakka])
+            (not (some (fn [u] (when (= (get-in app [:talvihoitoreitti :valittu-urakka :id]) (:urakka-id u)) true)) (:oikeudet-urakoihin app))))
```
Minä pimitän aina näistä, tässä alkaa olla yhdellä rivillä aika paljon logiikkaa, ei saa mitään selvää. Jos pari line breakkia iskisi tuohon

```clojure
(if (and 
	(get-in app [:talvihoitoreitti :valittu-urakka])
	(not 
	  (some (fn [u] 
			  (when (= (get-in app [:talvihoitoreitti :valittu-urakka :id]) (:urakka-id u)) 
				true)) 
		(:oikeudet-urakoihin app))))
```

rckopteri replied:
Laitoin pari rivivaihtoa

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,129 @@
+(ns harja.views.urakka.laadunseuranta.talvihoitoreitit-nakyma
+  "Sanktioiden ja bonusten välilehti"
```
Ei taida olla :D

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,129 @@
+            [harja.ui.debug :as debug]
+            [harja.ui.ikonit :as ikonit]
+            [harja.ui.komponentti :as komp]
+            [harja.asiakas.kommunikaatio :as k]))
```
Turhia:
harja.asiakas.kommunikaatio
ajax-loader
harja.ui.debug
reagent.core

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,129 @@
+              ;; Näytä valittu rivi kartalla tai piilota se
+              [:<>
+               (if (contains? valitut-kohteet (:id rivi))
+                 (harja.ui.napit/avaa "Piilota kartalta" #(e! (tiedot/->PoistaValittuKohdeKartalta (:id rivi))))
```
Importtaa napit

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta.cljs:
```diff
@@ -47,7 +49,10 @@
-    :mobiilityokalu (not (urakka/vesivaylaurakka? urakka))))
+    :mobiilityokalu (not (urakka/vesivaylaurakka? urakka))
+    :talvihoitoreititys (and (= :teiden-hoito tyyppi)
+                          (roolit/jvh? @istunto/kayttaja)
```
Onko tarkoitus  pitää vielä järjestelmänvastaavan roolissa?

rckopteri replied:
Ei ole. Hyvä bongaus!

solita-staskila reviewed (changes requested):
Koodinsiistintää

solita-staskila commented on src/cljs/harja/views/hallinta/tyokalut/talvihoitoreitit_tyokalu_nakyma.cljs:
```diff
@@ -0,0 +1,135 @@
+       {:nimi :tierekisteri-otsikko
+        :tyyppi :komponentti
+        :komponentti (fn [n]
+                       [:div "Voit syöttää kaksi eri tierekisteriosoitetta:"])}
```
Saiskohan tähän helposti kartan näkyviin mistä voisi valita pisteet? Hieman manuaalista hommaa, mutta näyttää toimivan kyllä propsit siitä

rckopteri replied:
Seuraavaksi aattelin panostaa vain listaukseen, josta näkisi kaikki ympäristössä käytössäolevat tierekisteriosoitteet. Kun näitä lähinnä tarvitaan tuolla GC tms koneilla. Noissa ympäristöissä on vaikea tietää, että millainen tieosoiteverkko siellä on kannassa olemassa. 
Eli kartalta valintakaan ei välttisti auta, kun tieosoitteita on kannassa niin vähän.

Mutta. Olen siis pohtinut tuotakin.

solita-staskila commented on src/cljc/harja/ui/kartta/asioiden_ulkoasu.cljc:
```diff
@@ -484,6 +484,12 @@ tr-ikoni {:img (pinni-ikoni "musta")
    {:color puhtaat/vihrea
     :width 6}])
 
+(def talvihoitoreitit
```
Nämä selitteet ei nyt täsmää mitä kartalla ja UI:lla näkyy. Tarviiko näitä näkyä? Designerille tästä kysymystä, koska näitä ei Figmassakaan ole. Mutta ei haluta väärääkään tietoa tuohon laittaa

rckopteri replied:
Hyvä pointti. Mulla on jäänyt tuo näimmä ihan kesken. Tarkoitus oli tietenkin näyttää noi samalla värillä, kuin mitä ne reititkin. Ja heitin designerillekin ideaa, että esim Traktorireitit tulisi eri värillä kuin Kuorma-autoreitit, mutta se ei vielä ottanut tuulta alleen.

Eli otin pois, kun ei niitä ole speksissäkään.

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,129 @@
+            (when (get talvihoitoreittien-tilat (:id rivi))
+              (when (> reittien-maara 0)
+                [:div {:style {:max-width "900px"}}
+                 [grid/grid
```
Tästä puuttuu otsikko "Reitti". 
![image](https://github.com/user-attachments/assets/e4837831-674e-46cc-9ece-3cec5bc2250b)

rckopteri replied:
oho. Niin puuttuukin. Laitoin.

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,129 @@
+                    :fmt #(fmt/desimaaliluku-opt % 2) :tasaa :oikea :leveys 2}
+                   {:otsikko "Laskettu pituus (km)" :nimi :laskettu_pituus :tyyppi :numero
+                    :fmt #(fmt/desimaaliluku-opt % 2) :tasaa :oikea :leveys 2}]
+                  (:reitit rivi)]]))])))]))
```
Lopusta puuttuu yhteensä sarake. 
![image](https://github.com/user-attachments/assets/eb37aa32-0e0a-48dd-9ff9-318a4020c761)

Vinkkinä: Gridille voidaan antaa :rivi-jalkeen-fn

rckopteri replied:
Jees.. Tämäkin jäänyt. 
Tuo on ulkoasultaan vähä erilainen meillä, kuin speksissä, mutta näillä on nyt mentävä.

solita-staskila replied:
https://github.com/finnishtransportagency/harja/pull/3513#discussion_r1776632720

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,129 @@
+             ;; Nimi
+             [:div.basis256.nogrow.shrink3.rajaus
+              [:div {:style {:display "flex"}}
+               [:span {:style {:margin-top "auto" :margin-bottom "auto" :margin-right "5px" :width "12px" :height "12px" :background-color (:vari rivi)}}]
```
Alkaa olla tyylittelyä sen verran, että tee näille oma lessi, ja heitä sinne.

rckopteri replied:
Siirsin

solita-staskila replied:
https://github.com/finnishtransportagency/harja/pull/3513#discussion_r1770814377

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,129 @@
+               (doall (for [h (:hoitoluokat rivi)]
+                        ^{:key (hash (str "hoitoluokka-" h))}
+                        [:div
+                         [:div.body-text.musta {:style {:padding-right "10px"}} (:hoitoluokka h)]
```
Kaikki nämä {:style .. } tässä koko tiedostossa voisi laittaa lessiin.
Tästä puuttuu bold-teksti. Otsikot designissa bold, ja kilometrit normal

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,129 @@
+               (doall (for [kalusto (:kalustot rivi)]
+                        ^{:key (hash kalusto)}
+                        [:div
+                         [:div.body-text.musta {:style {:padding-right "10px"}} (:kalustotyyppi kalusto)]
```
Kaipaisi hieman paddingia, sekä bold-tekstiä.
Vasemmalla nykyinen toteutus VS. oikealla design:
![image](https://github.com/user-attachments/assets/18c601f8-8cc7-4bb3-b31c-7707a1742a18)

rckopteri replied:
Ah. Speksejä on näimmä muutettu. Nuo oli määritelty weight 400:seksi. Muutin nyt semiboldiksi (weight 600) ja lisäsin vähä pädäriä oikealle.

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,129 @@
+             [:div.basis256.nogrow.shrink3.rajaus
+              [:div {:style {:display "flex"}}
+               [:span {:style {:margin-top "auto" :margin-bottom "auto" :margin-right "5px" :width "12px" :height "12px" :background-color (:vari rivi)}}]
+               [:div.semibold.musta.body-text (str (:nimi rivi))]]
```
Tämän alta puuttuu kokonaiskilometrit?
![image](https://github.com/user-attachments/assets/e87c28d7-e89e-481c-9a35-28b70610ff2c)

rckopteri replied:
Joo. Muutin tietomallia, niin yhteismäärää ei ollutkaan enää. Lisäilin tämän.

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,129 @@
+                   {:otsikko "Suunniteltu pituus (km)" :nimi :pituus :tyyppi :numero
+                    :fmt #(fmt/desimaaliluku-opt % 2) :tasaa :oikea :leveys 2}
+                   {:otsikko "Laskettu pituus (km)" :nimi :laskettu_pituus :tyyppi :numero
+                    :fmt #(fmt/desimaaliluku-opt % 2) :tasaa :oikea :leveys 2}]
```
Sarakkeet ovat leveämpiä designissa

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,129 @@
+              [:<>
+               (if (contains? valitut-kohteet (:id rivi))
+                 (harja.ui.napit/avaa "Piilota kartalta" #(e! (tiedot/->PoistaValittuKohdeKartalta (:id rivi))))
+                 (harja.ui.napit/avaa "Näytä kartalla" #(e! (tiedot/->LisaaValittuKohdeKartalle (:id rivi)))))]]]
```
Ikonit näissä napeissa on erilaisia designissa, sekä hieman eri keskitys, mutta minun  puolesta kyllä aivan hyvä näin.

solita-staskila reviewed (changes requested):
UI:hin sekä koodiin korjauksia, testattu toteumatyökalua, UI näkymää sekä Excel tuontia. Kaikki näyttää funktionaalisesti pelaavan ilman virheitä tai muita ongelmia, hienoa

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

rckopteri committed:
Yleistä koodin refaktorointia ja waroitusten poistoa

rckopteri committed:
Paranna Talvihoitoreittien exceltuonnin validointia

Ota käyttöön ja tee yleiseksi validointifunktioksi olemassa oleva tieosoitteen validointifunktio

rckopteri committed:
Refaktorointia katselmointikommenttien perusteella

rckopteri committed:
Hallintapaneeliin mahdollisuus katsoa käytössä olevan ympäristön tieosoitteita tietokannasta

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,137 @@
+         :lataus-epaonnistui #(e! (tiedot/->TiedostoLadattu %))
+         :tiedosto-ladattu #(e! (tiedot/->TiedostoLadattu %))}]
+       [yleiset/tiedoston-lataus-linkki
+        "Lataa Excel-pohja"
```
Saisiko tähän väliin hieman paddingia laitettua? 
Jos kartta piilotettuna, tulee pientä päällekkäisyyttä

![image](https://github.com/user-attachments/assets/8821f9ff-fe02-421c-8dd6-1f7c79f60cbb)

rckopteri replied:
Tuo on tosi tympeä tuo karttanappula. Se on kaikilla sivuilla samassa paikassa. En lähde tämän fiitserin puitteissa sitä muokkaamaan.

solita-staskila replied:
Tuon featuren päänäkymään voi lisätä vähän paddingia päälle, minkä tein reikäpaikkauksissa. Mutta OK.

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,137 @@
+                   {:otsikko "Hoitoluokka" :nimi :hoitoluokka :tyyppi :string :tasaa :vasen :leveys 2}
+                   {:otsikko "Suunniteltu pituus (km)" :nimi :pituus :tyyppi :numero
+                    :fmt #(fmt/desimaaliluku-opt % 2) :tasaa :oikea :leveys 2}
+                   {:otsikko "Laskettu pituus (km)" :nimi :laskettu_pituus :tyyppi :numero
```
> Jees.. Tämäkin jäänyt.
> Tuo on ulkoasultaan vähä erilainen meillä, kuin speksissä, mutta näillä on nyt mentävä.

Tuon saapi helposti, tai kun itse olen näitä tunkannut niin paljon :D 
Huom Figman taustavärit (pure white), ero todella pieni, mutta on olemassa

Nykyinen toteutus:
![image](https://github.com/user-attachments/assets/dcb4e386-029a-47d6-8156-9c21c9343188)


Tältä pitäisi näyttää: 
![image](https://github.com/user-attachments/assets/cb23ba8d-80e7-42d4-a411-0cc6eec28438)

Oisko vaikka näin!

```clojure
[grid/grid
  {:salli-valiotsikoiden-piilotus? true
   :valiotsikoiden-alkutila :kaikki-kiinni
   :tunniste :id
   :reunaviiva? true
   :rivi-jalkeen-fn #(let [yhteensa-suunniteltu (reduce + 0 (map :pituus %))
						   yhteensa-laskettu (reduce + 0 (map :laskettu_pituus %))]
					   [{:teksti "" :luokka "otsikko-ei-taustaa"}
						{:teksti "Yhteensä" :luokka "lihavoitu otsikko-ei-taustaa"}
						{:teksti "" :luokka "otsikko-ei-taustaa"}
						{:teksti (str (fmt/euro-opt false yhteensa-suunniteltu))
						 :tasaa :oikea :luokka "lihavoitu otsikko-ei-taustaa"}
						{:teksti (str (fmt/euro-opt false yhteensa-laskettu))
						 :tasaa :oikea :luokka "lihavoitu otsikko-ei-taustaa"}])}

  [{:otsikko "Tie" :luokka "nakyma-valkoinen-solu" :nimi :tie :tyyppi :string :tasaa :vasen :leveys 1}
   {:otsikko "Tieosoite" :luokka "nakyma-valkoinen-solu" :nimi :formatoitu-tr :tyyppi :string :tasaa :vasen :leveys 3}
   {:otsikko "Hoitoluokka" :luokka "nakyma-valkoinen-solu" :nimi :hoitoluokka :tyyppi :string :tasaa :vasen :leveys 2}
   {:otsikko "Suunniteltu pituus (km)" :luokka "nakyma-valkoinen-solu" :nimi :pituus :tyyppi :numero
	:fmt #(fmt/desimaaliluku-opt % 2) :tasaa :oikea :leveys 2}
   {:otsikko "Laskettu pituus (km)" :luokka "nakyma-valkoinen-solu" :nimi :laskettu_pituus :tyyppi :numero
	:fmt #(fmt/desimaaliluku-opt % 2) :tasaa :oikea :leveys 2}]
  (:reitit rivi)]
```

Voit copy pastettaa suoraan

rckopteri replied:
Okei. Täähän oli näppärä. Oletin, että meillä on gridi noissa perustaulukoissa ihan valmis paketista.

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,137 @@
+
+             ;; Nimi
+             [:div.basis256.nogrow.shrink3.rajaus
+              [:div {:style {:display "flex"}}
```
Tässä tiedostossa vieläkin tosi paljon näitä `:style` tyylityksiä.. Kun tehtiin jo se oma lessi, nämä kaikki voisi heittää minusta sinne. Ymmärrän että tämä voi viedä vähän aikaa, mutta minä suosin lessiä enemmän kun että kirjoitetaan clojureen, kun clojure on  muutenkin jo niin mössöä :D Mutta tee kuten näät parhaaksi

rckopteri replied:
Otin mallia noihin muualta ja minusta tuon voisi kyllä tehdä lessillä näppärämmin. Mutta helpompi ottaa mallia jo olemassaolevasta, kuin rakentaa uusiksi lessillä, niin jätän se näin.

solita-staskila commented on src/cljs/harja/tiedot/urakka/laadunseuranta/talvihoitoreitit_tiedot.cljs:
```diff
@@ -0,0 +1,227 @@
+      (reset! nav/kartan-extent alue)
+      app))
+
+  TiedostoLadattu
```
Olisko se Ajax loaderin lisäys tännekin tarpeellinen?
Tämä on niin helppoakin, eli täällä tiedot.cljs, kun aloitetaan operaatio, palautat vaan:
`(assoc app :haku-kaynnissa? true)`,  ja nytkun katson, tämä on jo täälläkin olemassa.

Eli  fronttiin (viewssiin) nyt vaan:
```clojure
(if haku-kaynnissa?
       [:div.ajax-loader-valistys
        [ajax-loader-pieni (str "Haetaan tietoja...")]] 
.. else  näytä grid
)
```

rckopteri replied:
Wrappäsin vielä tuo ajaxloaderin tuonne näkymään matkaan. Se vilahtaa niin äkkiä, ettei sitä kukaan ehdi huomata, mutta ehkä se auttaa silti.

rckopteri replied:
Eniten tarvitsisi kahvaa siihen tiedoston käsittelyyn, mutta en tällä haavaa lähde sitä tutkimaan. Siihen ei ole ihan niin simppeliä pultata latausikonia kiinni.

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta/talvihoitoreitit_nakyma.cljs:
```diff
@@ -0,0 +1,137 @@
+         (kartta-tasot/taso-paalle! :talvihoitoreitit)
+         (kartta-tasot/taso-paalle! :organisaatio)
+         (reset! tiedot/karttataso-nakyvissa? true)
+         (js/console.log "sisään"))
```
Delete

rckopteri replied:
otin pois

solita-staskila commented on src/cljs/harja/tiedot/urakka/laadunseuranta/talvihoitoreitit_tiedot.cljs:
```diff
@@ -0,0 +1,227 @@
+
+  TiedostoLadattu
+  (process-event [{:keys [vastaus]} app]
+    (let [_ (js/console.log "TiedostoLadattu :: vastaus:" (pr-str vastaus))
```
Delete tämä printti

ps. katso aikaisempi kommentti (tämän rivin alla olevat filtterit):

> Tässä muut-virheet muuttujassa nyt tuon filtterin sisällä jotain logiikan kanssa väärin? Verrataan = operaattorilla itseään. Eli onko tuo aina true? Jotain tässä nyt on väärin, tai muuten vaan turhaa koodia. En ole tätä testannut kirjoitushetkellä, luen pelkästään koodia läpi

https://github.com/finnishtransportagency/harja/pull/3513#discussion_r1770787072

solita-staskila commented on src/cljs/harja/tiedot/hallinta/tyokalut/talvihoitoreitti_tyokalu.cljs:
```diff
@@ -0,0 +1,240 @@
+
+  LisaaKirjoitusOikeusEpaonnistui
+  (process-event [{vastaus :vastaus} app]
+    (js/console.log "LisaaKirjoitusOikeusEpaonnistui :: vastaus" (pr-str vastaus))
```
Voi deletoida kaikki nämä printit? Näitä tässä tiedostossa paljon

solita-staskila commented on src/clj/harja/palvelin/palvelut/laadunseuranta/talvihoitoreitit_palvelu.clj:
```diff
@@ -0,0 +1,203 @@
+                             nil))
+
+        ;; Käsittele jokainen talvihoitoreitti itsenäisessä loopissa
+        _ (dorun (for [t talvihoitoreitit]
```
Oliko tässä niitä muisti ongelmia? Pääsitkö tässä eteenpäin? Jos sinulla on se testi exceli jossa  tämä virhe esiintyy, minä voin tätä myös yrittää korjata.

rckopteri replied:
Jep. Tuo for silmukan käyttö ratkaisi ongelman. se doseq oli muistisyöppö

solita-staskila commented on resources/api/schemas/talvihoitoreitti-kirjaus-request.schema.json:
```diff
@@ -0,0 +1,104 @@
+      }
+    },
+    "required": [
+      "ulkoinen-id",
```
`ulkoinen-id` on required, mutta tämä puuttuu skeemasta??

rckopteri replied:
Hyvä huomio. Oikea nimi on tunniste

solita-staskila commented on resources/api/schemas/talvihoitoreitti-kirjaus-request.schema.json:
```diff
@@ -0,0 +1,104 @@
+                  "description": "Kaluston lukumäärä, jota kalustotyypistä käytetään. Esim. 4",
+                  "minimum": 1
+                },
+                "required": ["kalustotyyppi", "kalusto-lkm"]
```
Onko tämä nyt jotenkin nestattu require? 
Onko tämä validi? Eikö nämä requiret pitäisi declarettaa tuossa properties objectin tasolla?
Minusta tämä on nyt yhden tason liian syvällä

Eli...

```json
"kalustot": 
{
  ..
  "items": 
  {
    ..
    "properties": 
    {
      "kalustotyyppi": {
        ..
      },
      "kalusto-lkm": {
        ...
      }
    }, // <- "properties"
    "required": ["kalustotyyppi", "kalusto-lkm"]
  }
}
```

rckopteri replied:
Totta! hyvä huomio.

solita-staskila commented on resources/api/schemas/talvihoitoreitti-kirjaus-request.schema.json:
```diff
@@ -0,0 +1,104 @@
+            "required": true,
+            "description": "Reitin hoitoluokka. Hoitoluokkavaihtoehdot: IsE, Is, I, Ib, Ic, II, III, L, K1, K2, K"
+          },
+          "kalustot": {
```
Onko tarvetta  määrittää millekään näistä `minItems:"?"` Tarkista

rckopteri replied:
Sen excelin mukaan sieltä voi tulla esim 1 TR, 2 KA ja 1 KUP, tai sitten vain 1 TR. Eli ei ole tiedossa tarkkoja minimiarvoja. Mutta kalustot on pakollisia ja kun kalustoa syöttää niitä lukumäärä ja kalustotyyppi on pakollisia.

solita-staskila commented on resources/api/schemas/talvihoitoreitti-paivitys-request.schema.json:
```diff
@@ -0,0 +1,108 @@
+                },
+                "required": [
+                  "kalustotyyppi",
+                  "kalusto-lkm"
```
Onko nämäkin nyt väärällä tasolla? Pitäisikö nämä nakkaa tuonne alemmalle tasolle?

```json
"kalustot": 
{
  "type": "array",
  "items": 
  {
    "type": "object",
    "properties": 
    {
      "kalustotyyppi": {
        ..
      },
      "kalusto-lkm": {
        ..
      }
    },
    "required": ["kalustotyyppi", "kalusto-lkm"]
  }
}
```

solita-staskila commented on resources/api/schemas/talvihoitoreitti-paivitys-request.schema.json:
```diff
@@ -0,0 +1,108 @@
+                },
+                "kalusto-lkm": {
+                  "type": "integer",
+                  "required": true,
```
Tekeeköhän tämä require true nyt mitään? Kun meillä on jo tuo erillinen required definition olemassa. 

Jos meillä on jo:` "required": ["kalustotyyppi", "kalusto-lkm"]`
Niin miksi pitää olla vielä `required: true` tämän itemin sisällä?

Täältä esim:
https://davidwalsh.name/json-validation

Katso rvaneperen kommentti:

> I tried "serverEnvironmentId": { "type": "number", "required" : true }, but it still passed the test even though that element does not exist in the JSON data.

Tiedän, että näitä `required: true` on järjestelmä täynnä, mutta nyt täytyisi tarkistaa tekeekö tuo oikeasti mitään, koska minä en usko. Pystytkö tarkistamaan tämän?

Eli jos poistat `"required": ["kalustotyyppi", "kalusto-lkm"]`, ja jätät vaan tuon `required:true`, saatko chekattua toimiiko tämä?

Ja  myös  toisinpäin, onko mitään eroa, jos on pelkästään:
`"required": ["kalustotyyppi", "kalusto-lkm"]`
vs.
`required:true`

Kiitos ja anteeksi, en halua teettää työtä mutta tämä on nyt aivan selkeä epäyhtenäisyys meidän skeemoissa

solita-staskila commented on resources/api/schemas/talvihoitoreitti-poisto-request.schema.json:
```diff
@@ -0,0 +1,17 @@
+      "type": "array",
+      "minItems": 1,
+      "maxItems": 1000,
+      "required": true,
```
Täälläkin. Olisko nyt vaan näin:

```json
{
  "$schema": "http://json-schema.org/draft-03/schema#",
  ...
  "properties": {
    "talvihoitoreittien-tunnisteet": {
      ...
    }
  },
  "required": ["talvihoitoreittien-tunnisteet"]
}
```

Näille pitäisi jokin yhtenäisyys saada.

rckopteri replied:
Tutkailin noita määrityksiä skeemojen versioille. Propertien alla "required": true on draft-03 tyyli. Ja string array property listan lopussa on draft-04:Sta alkaen. Meillä on pääosin käytössä draft-03 (joka ollaan varmaan vaihtamassa vielä jossain vaiheessa), joten otin array listat pois.

solita-staskila commented on dev-resources/less/pages/talvihoitoreititys.less:
```diff
@@ -0,0 +1,7 @@
+  margin-bottom: auto;
+  margin-right: 5px;
+  width: 12px;
+  height: 12px;
```
Käytä `@valistys-12`

solita-staskila commented on dev-resources/less/pages/talvihoitoreititys.less:
```diff
@@ -0,0 +1,7 @@
+.talvihoitoreitti-nimi {
+  margin-top: auto;
+  margin-bottom: auto;
+  margin-right: 5px;
```
Käytä mielummin `@valistys-4`, ellei tuo 1 pixeli sitten heitä jotain vinoon.

solita-staskila reviewed (changes requested):
Uusi tuomionvasara, ole hyvä 🔨

UI näyttää nyt aika hyvältä! Mutta skeemoista annan vielä punasta valoa.
Huomioi nuo githubin piilottamat kommentit. Ja kaikki aikaisemmat auki olevat kommentit myös, laitoin resolved kaikkiin mitkä on OK, mutta ne jotka on auki, on vielä auki.

rckopteri committed:
Päivitä tuonti excel pohja talvihoitoreiteille

rckopteri committed:
UI refaktorointia

rckopteri committed:
Lisätään parille tielle vielä tieosoite

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

rckopteri committed:
Lisää ramli tiedostoon tiedot talvihoitoreittien apikutsuista

rckopteri committed:
Päivitä talvihoitoreittien yksikkötestit

rckopteri committed:
Lisää talvihoitoreitti apille tarkemman palvelun puolella olleet validoinnit

rckopteri committed:
Refaktorointia katselmointikommenttien perusteella

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

rckopteri committed:
Excelin rivinumerointiin ei voi ryhmittelyn ja validoinnin jälkeen enää luottaa, joten oteteaan se pois

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

rckopteri committed:
Kavenna talvihoitoreitin kartan piirron viivaa, aseta maksimileveys näytä ja piilota kartta napeille

rckopteri committed:
Tallenna värikoodi talvihoitoreitille lisäysvaiheessa, niin se pysyy samana aina, kun sivu avataan

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

# Conflicts:
#	resources/api/api.raml
#	src/clj/harja/kyselyt/urakat.clj
#	src/clj/harja/palvelin/main.clj
#	src/cljc/harja/domain/oikeudet.cljc
#	test/clj/harja/palvelin/main_test.clj
#	tietokanta/src/main/resources/db/migration/V1_1135__.sql

rckopteri committed:
Korjataan migraatiossa tapahtuneita virheitä

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1146__.sql

rckopteri committed:
Käytä eri hakua tieosoitteen oikeellisuuden todentamiseksi.

En löytänyt syytä miksi tr_osoitteet taulussa ei ole kaikki tieosoitteet, mutta tr_osien_pituudet taulussa on.

rckopteri committed:
Paranna talvihoitoreittien tallentamisen virhelogitusta

rckopteri committed:
Lue talvihoitoreittien excelistä reitin pituus sekä metreinä, että kilometreinä.

rckopteri committed:
Päivitä talvihoitoluokkien nimet Ise:ksi joka on oikea

rckopteri committed:
Päivitä ohjeita talvihoitoluokkien sisäänajolle lokaalisti

rckopteri committed:
Lisätään kolme uutta talvihoitoluokkaa. Hoitoaukot ja pysäköintialueet

rckopteri committed:
Päivitä talvihoitoreitin APIn toimintaa ja varmista se testein

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

# Conflicts:
#	resources/api/api.raml
#	tietokanta/src/main/resources/db/migration/V1_1147__.sql

rckopteri committed:
Korjaa talvihoitoreittien api.ramlia

rckopteri committed:
Lisää api.ramliin talvihoitoreitin poisto esitys

rckopteri committed:
Tarkenna testeja koodimuutosten jälkeen

rckopteri committed:
Anna oikeudet vain jvh käyttäjälle Talvihoitoreitteihin

rckopteri committed:
Päivitä ohjetuontiexceliä talvihoitoreiteille.

Lisätään kolme puuttuvaa talvihoitoluokkaa ja muutetaan pituus kilometreiksi

rckopteri committed:
Nimeä excelistä saatavat uudet talvihoitoreitit koneen ymmärtämään muotoon.

rckopteri committed:
Korjaa tyhjän excelin virheilmoitus

rckopteri committed:
Talvihoitoreiteille talvihoitoluokkien ryhmittely

rckopteri committed:
Varmistusta excelistä luvun pituuden käsittelyyn.

rckopteri committed:
Korjaa löytyneitä koodivirheitä

rckopteri committed:
Päivitä testit vastaamaan muuttunutta toteutusta

rckopteri committed:
Korjaa cypress testit vastaamaan muuttunutta toteutusta.

rckopteri committed:
Laita talteen konsti korjata pikaisesti lokaaliympäristö sopivaksi cypress -testejä varten.

Kunnes kirjastoon liittyvät muutokset on tehty.

solita-staskila committed:
Wrappaa rivit päägridin sisälle sivutuksella

solita-staskila committed:
Tee riville oma funktio

solita-staskila committed:
Siirrä less luokat omaan tiedostoon

solita-staskila commented on tietokanta/src/main/resources/db/migration/V1_1148__.sql:
```diff
@@ -0,0 +1,45 @@
+-- Talvihoitoreitityksille tietokantataulut
+CREATE TABLE IF NOT EXISTS talvihoitoreitti
```
Olisko tarve, jos ei nyt  niin joskus, että taululla olisi poistettu?-  sarake?

rckopteri replied:
Jes. Hyvä pointti. Tämä tehdään erillisellä pr:llä sitten, kun fiitseri on saatu tuotantoon.

solita-staskila commented on src/clj/harja/kyselyt/talvihoitoreitit.sql:
```diff
@@ -0,0 +1,92 @@
+VALUES (:sijainti_id, :kalustotyyppi, :maara);
+
+
+-- name: hae-urakan-talvihoitoreitit
```
Tulokset ei oo missään järjestyksessä ja pukkaa randomilla, tarviiko olla? 
Gridillekin voi nyt passata `jarjesta :id` esim.

rckopteri replied:
sql ruukaa palauttaa id järjestysksessä, jos orderia ei ole annettu. Mutta lisäsin siitä huolimatta. Hyvä nosto.

solita-staskila commented on src/clj/harja/palvelin/palvelut/laadunseuranta/talvihoitoreitit_excel.clj:
```diff
@@ -0,0 +1,139 @@
+        ;; Reitti ja kalusto tabeilla täytyy olla sama määrä rivejä
+        _ (when (not= (count reitti-rivit) (count kalusto-rivit))
+            (throw+ {:type :validaatiovirhe
+                     :virheet [{:virheet (str "Reittien ja kaluston määrä ei täsmää. Reittien määrä: " (count reitti-rivit)
```
Kuinka tän kuuluis nyt toimia, kun tää nyt sanoo että reittien määrä 0? Onks mulla tässä jotain väärin? 

&nbsp;
<details>
<summary> ᴄʟɪᴄᴋ ᴛᴏ ᴇxᴘᴀɴᴅ </summary>
&nbsp;

![Screenshot 2024-11-12 175126](https://github.com/user-attachments/assets/10044d87-da98-4a78-9f8d-6a2c766cd385)

![Screenshot 2024-11-12 175136](https://github.com/user-attachments/assets/d7659d36-2db0-4650-8511-d9defd7581c1)

![Screenshot 2024-11-12 175148](https://github.com/user-attachments/assets/6f4b5853-934b-4f84-b77f-b5c674e863de)

![Screenshot 2024-11-12 175312](https://github.com/user-attachments/assets/25225822-3fc0-4c36-b86f-733c06e3b20a)

![Screenshot 2024-11-12 175421](https://github.com/user-attachments/assets/02a83248-8a01-44c3-9876-21d9f922eb2a)


![Screenshot 2024-11-12 175440](https://github.com/user-attachments/assets/47884207-adb8-4d89-8cb6-b6f5e893bee5)






</details>

rckopteri replied:
Siinä excelissä on 2 välilehteä. Toisessa määritellään kalusto, toisessa tieosoitteet. Sulla on varmaan siellä toisella välilehdellä jotain väärin, jotta niiden määrät ei täsmää.

solita-staskila replied:
Joo, en ollut sinne laittanut mitään, itse  vaan olin sokea  👨‍🦯

solita-staskila commented on src/cljc/harja/ui/kartta/infopaneelin_sisalto.cljc:
```diff
@@ -843,6 +843,13 @@
             {:otsikko "Aikataulu" :tyyppi :string :nimi :aikataulu}]
    :data (:infopaneelin-tiedot kohde)})
 
+(defmethod infopaneeli-skeema :talvihoitoreitit [talvihoitoreitti]
```
Tulee vähän warningia kun klikkaa karttaa. En muista mikä tässä oli kuvio.

```
Tiedolla ei ole yhtään validia skeemaa: {:tyyppi :talvihoitoreitit ... } 
```

![image](https://github.com/user-attachments/assets/df5c686b-d02b-4192-bf69-44d68a4e9364)

rckopteri replied:
Voikun ikävä. Se infopaneelidata on näimmä annettava. Muuten tulee aina tuo virhe. Joten lisäsin sen infopaneelin sinne takaisin.

solita-staskila commented on src/cljs/harja/views/urakka/laadunseuranta.cljs:
```diff
@@ -47,7 +49,11 @@
+    :talvihoitoreititys (and
+                          (= :teiden-hoito tyyppi)
+                          (oikeudet/urakat-laadunseuranta-talvihoitoreititys id)
+                          (roolit/jvh? @istunto/kayttaja))))
```
Jätetäänkö?

rckopteri replied:
Otin tästä vielä pois tuon koko osiovaatimuksen. Eli jätin pelkästään jvh:lle oikeudet.

solita-staskila commented on project.clj:
```diff
@@ -17,6 +17,8 @@
 
                  [prismatic/schema "1.4.1"]
+                 ; Käytä lokaalisti
+                 ;[prismatic/schema "1.1.10"]
```
Tää oli ilmeisesti Cypressejä varten. Muista ottaa tää pois ennen mergeä, voi olla vaarallinen

rckopteri replied:
Joo. Otan pois ennen mergeä.

solita-staskila reviewed (approved):
Tein tuoun sivutuksen mistä puhuimma, ja siirsin lessi luokat lessiin. 
Vaihdoin samalla kartan koon **:L** -> **:M**, minusta tuo **:L** on vakiona tähän liian iso. 
Voi aiheutua Cypress faileja, pahoittelut siitä.

Kato kommentit tarkoin, yks rivi on mitä ei haluta viedä tuotantoon.
Hyvää jobia

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

# Conflicts:
#	project.clj
#	resources/roolit.xlsx
#	src/clj/harja/palvelin/main.clj
#	test/clj/harja/palvelin/main_test.clj
#	tietokanta/src/main/resources/db/migration/V1_1148__.sql

rckopteri committed:
Refaktorointia katselmointikommenttien perusteella.

rckopteri committed:
Merge branch 'develop' into HARJA-9-talvihoitoreitit

rckopteri committed:
Poista turhia logituksia ja cljs require virhe

rckopteri committed:
Päivitetään cy testit vastaamaan muuttunutta toteutusta

rckopteri committed:
Lisää konfliktissa kadonnut komponentti

rckopteri committed:
Pituuden laskennasta ihan tarkka vertailu pois
